Automatic task-based concurrency using local streams

maleadt opened the pull request:
This is the next big step (after https://github.com/JuliaGPU/CUDAnative.jl/pull/609 and https://github.com/JuliaGPU/CUDA.jl/pull/395) in marrying Julia's task-based parallelism with the CUDA APIs. By managing our own default stream, and using that instead of CUDA's default stream objects, we can ensure that all operations (kernel launches, library calls, etc) happen using the stream that's active for the current task. That should make it much easier to isolate independent operations. For example:

```julia
# operation that both executes a kernel and calls a library function
julia> run(a,b,c) = (mul!(c, a, b); broadcast!(sin, c))

julia> function main(N=1024)
       a = CUDA.rand(N,N)
       b = CUDA.rand(N,N)
       c = CUDA.rand(N,N)
       NVTX.@range "warmup" run(a,b,c)
       synchronize()
       NVTX.@range "global" x = run(a,b,c)
       NVTX.@range "local" y = CUDA.stream!(CuStream()) do
           run(a,b,c)
       end
       synchronize()
       x == y
       end
```

![image](https://user-images.githubusercontent.com/383068/104914100-ebba1100-598e-11eb-98bd-1e50f41c8dc6.png)

As you can see, the operations in the global stream are independent from the ones executed in a local stream context :tada: Here, that results in overlapping execution, which is great for performance.

Remains to be done/decided:
- should `synchronize` now default to synchronizing the current stream, or should it still be a device-wide sync?
- we can further improve this by using CUDA 11.2's async allocation features.

maleadt committed:
Manage our own per-task stream.

maleadt commented:
Changing the stream on a task now updates the active task-bound handles of libraries like CUBLAS and CUDNN. That means switching streams isn't entirely free; so it might be costly to write code that performs a single operation like that. That doesn't seem very realistic though, so I don't think it's worth additional complexity (like lazily changing the handle's stream).

maleadt committed:
Support for async alloc/free.

maleadt commented:
Current overhead:

```
julia> @benchmark @cuda identity(nothing)
BenchmarkTools.Trial: 
  memory estimate:  400 bytes
  allocs estimate:  10
  --------------
  minimum time:     1.339 μs (0.00% GC)
  median time:      1.383 μs (0.00% GC)
  mean time:        1.403 μs (0.00% GC)
  maximum time:     4.159 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> @benchmark CUDA.stream!(()->@cuda(identity(nothing)), s) setup=(s=CUDA.stream())
BenchmarkTools.Trial: 
  memory estimate:  1.36 KiB
  allocs estimate:  53
  --------------
  minimum time:     2.517 μs (0.00% GC)
  median time:      2.606 μs (0.00% GC)
  mean time:        2.756 μs (2.62% GC)
  maximum time:     367.082 μs (98.34% GC)
  --------------
  samples:          10000
  evals/sample:     9
```

maleadt commented:
```
julia> @benchmark @cuda identity(nothing)
BenchmarkTools.Trial: 
  memory estimate:  400 bytes
  allocs estimate:  10
  --------------
  minimum time:     1.300 μs (0.00% GC)
  median time:      1.343 μs (0.00% GC)
  mean time:        1.364 μs (0.00% GC)
  maximum time:     3.986 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> @benchmark CUDA.stream!(()->@cuda(identity(nothing)), s) setup=(s=CUDA.stream())
BenchmarkTools.Trial: 
  memory estimate:  400 bytes
  allocs estimate:  10
  --------------
  minimum time:     1.421 μs (0.00% GC)
  median time:      1.477 μs (0.00% GC)
  mean time:        1.493 μs (0.00% GC)
  maximum time:     4.049 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> # let's initialize some libraries

julia> CUBLAS.handle(); CURAND.default_rng(); CUSPARSE.handle(); CUSOLVER.dense_handle(); CUSOLVER.sparse_handle();

julia> @benchmark CUDA.stream!(()->@cuda(identity(nothing)), s) setup=(s=CUDA.stream())
BenchmarkTools.Trial: 
  memory estimate:  400 bytes
  allocs estimate:  10
  --------------
  minimum time:     1.452 μs (0.00% GC)
  median time:      1.531 μs (0.00% GC)
  mean time:        1.543 μs (0.00% GC)
  maximum time:     8.312 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10
```

maleadt committed:
Re-use the library handle caches to lower stream switching overhead.

vchuravy commented:
> Changing the stream on a task now updates the active task-bound handles of libraries like CUBLAS and CUDNN. That means switching streams isn't entirely free; so it might be costly to write code that performs a single operation like that. That doesn't seem very realistic though, so I don't think it's worth additional complexity (like lazily changing the handle's stream).

If we keep the stream as a keyword argument to low-level functions I agree, for KA purposes I need to run a short sequence of ops on the same stream and then restore the previous one, or I need to pass a explicit stream object for `async_copy`, `record`, `wait` and `launch`.

vchuravy commented on lib/cudadrv/events.jl:
```diff
@@ -37,12 +37,11 @@ Base.:(==)(a::CuEvent, b::CuEvent) = a.handle == b.handle
+    record(e::CuEvent)
 
-Record an event on a stream.
+Record an event on the currently-acttive stream.
```
```suggestion
Record an event on the currently-active stream.
```

vchuravy commented on lib/cudadrv/memory.jl:
```diff
@@ -62,18 +62,26 @@ Allocate `bytesize` bytes of memory on the device. This memory is only accessibl
 for access on the CPU.
 """
-function alloc(::Type{DeviceBuffer}, bytesize::Integer)
+function alloc(::Type{DeviceBuffer}, bytesize::Integer; async::Bool=false)
```
Should we make all allocation async by-default?

maleadt replied:
CUDA 11.2-only... I've thought about bumping the minimum requirement, but it might be a bit much.

vchuravy commented on lib/cudadrv/memory.jl:
```diff
@@ -241,7 +249,7 @@ end
 """
 function prefetch(buf::UnifiedBuffer, bytes::Integer=sizeof(buf);
-                  device::CuDevice=CuCurrentDevice(), stream::CuStream=CuDefaultStream())
+                  device::CuDevice=device(), stream::CuStream=stream())
```
Still has a stream kw-arg

maleadt replied:
Hm, I guess we should remove the `device` argument there too, then? Or maybe that's pushing it too far... and then we should probably retain most of the `stream` arguments too...

maleadt commented:
> for KA purposes I need to run a short sequence of ops on the same stream and then restore the previous one

Is the switching overhead low enough now? We could add some stream kwargs, but again where to draw the line. I imagine I'd have to add most of the ones removed in this PR back. I could imagine adding it back for kernel launches though, as `@cuda dynamic=true` still has the `stream` argument anyway. And for other functions its easier to use the lower-level API (crafting a kernel launch like that is pretty tricky).

vchuravy reviewed (approved):
I like this direction a whole lot! I would push for making most memory operations use the async variant by default (iirc the non-async are device level syncs) and then emit a `sync` on the stream for `async=false`, i.e. have `async` represent the semantics w.r.t. the host. KA uses a busy-wait https://github.com/JuliaGPU/KernelAbstractions.jl/blob/c9c5daa579653877cd9f7893633d4fc224881cce/src/backends/cuda.jl#L63-L72 but that might not be worth it.

I will open a corresponding PR to KA to test these changes there.

cc: @alir @simonbyrne @kpamnany @jakebolewski

maleadt commented:
> I would push for making most memory operations use the async variant by default (iirc the non-async are device level syncs) and then emit a `sync` on the stream for `async=false`, i.e. have `async` represent the semantics w.r.t. the host.

That's interesting. That way, the `stream` argument actually means something in the `async=false` case (with regular CUDA APIs, you can't specify the stream to sync). That would remove most of the ugliness that the `stream` kwarg has, so maybe we should keep it then.

maleadt committed:
Remove some stream arguments.

maleadt committed:
Use asynchronous APIs for memory operations.

Have them synchronize only the currently-active stream, if requested.

maleadt committed:
Clean-up stream getter interface.

maleadt committed:
Properly handle global streams in the context of multiple devices.

maleadt committed:
Try out unique stream per task.

This makes task overlap by default, and means we don't have to
bother with the default stream semantics (ptsz/ptds).

maleadt committed:
Use a thread-local cache for the current stream.

Reduces the cost of querying the stream from ~20 to 4ns.

maleadt commented:
Update: I retained stream arguments to most functions for the purpose of KA.jl. At the same time though, the performance overhead of querying/switching streams has been greatly reduced, so the new API could be used too. I've also pushed a commit that assigns each task its own non-blocking stream, so tasks will overlap their computations automatically (I would have expected this to break something at least, but apparently we don't accidentally rely on default stream semantics). And memory operations now always use the async APIs + synchronize only the current stream (not the whole device), but still default to synchronous behavior.

maleadt committed:
Remove per-thread streams.

We use a task-local stream now, so don't need these default stream semantics.

maleadt committed:
Remove use of ptds/ptsz APIs.

Since we now use explicit per-task streams, we don't rely on
these default stream semantics anymore.

maleadt committed:
Update tests.

maleadt committed:
Make synchronize() only operate on the task-local stream.

maleadt committed:
Make host to device copies asynchronous by default.

Async H2D copies require pinned memory, but if it isn't
the copy will just execute synchronously.

maleadt committed:
Handle blocking during synchronization ourselves.

That way we can yield to other tasks.

vchuravy commented on lib/cudadrv/stream.jl:
```diff
@@ -98,6 +99,19 @@ function query(s::CuStream)
+Wait until a stream's tasks are completed. If `blocking` is true (the default), Julia will
+be asked to yield to any other scheduled task.
+"""
+function synchronize(s::CuStream=stream(); blocking::Bool=true)
```
`blocking` is a bit of an odd name. The function is blocking nonetheless.

maleadt replied:
You're right; It's because of cudaEventBlockingSync that I chose this name: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1g26509a522be9d449aa7c8c279612452d
> cudaEventBlockingSync: Specifies that event should use blocking synchronization. A host thread that uses cudaEventSynchronize() to wait on an event created with this flag will block until the event actually completes.

maleadt replied:
Let's keep it `blocking` to mirror CUDA, unless you feel strongly about this?

maleadt commented:
Well, this is promising. By handling blocking ourselves now during `synchronize`, and yielding back to Julia's task scheduler, we can naively sync from tasks without global effects:

```julia
using CUDA, LinearAlgebra

# dummy calculation (that does not allocate or otherwise synchronize the GPU)
function run(a,b,c)
    NVTX.@range "mul!" mul!(c, a, b)  # uses CUBLAS, so needs a library handle
    NVTX.@range "broadcast!" broadcast!(sin, c, c)
end

# one "iteration", performing the above calculation twice in two tasks
# and comparing the output.
function iteration(a,b,c)
    x,y = missing, missing
    NVTX.@range "iteration" @sync begin
        @async begin
            x = NVTX.@range "run 1" run(a,b,c)
            synchronize()
        end
        @async begin
            y = NVTX.@range "run 2" run(a,b,c)
            synchronize()
        end
    end
    # no need to synchronize here, as both tasks have been synchronized already
    x == y
end

function main(N=1024)
    a = CUDA.rand(N,N)
    b = CUDA.rand(N,N)
    c = CUDA.rand(N,N)
    synchronize() # to make sure we can use this data from other tasks
    NVTX.@range "warmup" iteration(a,b,c)
    GC.gc(true) # we want to collect and cache the library handles used during warmup
    NVTX.@range "main" iteration(a,b,c)
end
```

With some improvements in caching library handles (turns out creating them is expensive, so we can't naively do so each time a new task performs its first library call), we get very nice concurrent execution from the above example (which doesn't even use unreasonably-large inputs to hide latency):

![image](https://user-images.githubusercontent.com/383068/105473320-48724000-5c9d-11eb-8429-785edcfb49d5.png)

The tiny marks at the bottom are calls to `yield()`. As you can see at the top of the trace, execution of independent `mul!` and `broadcast!` are overlapping.

maleadt committed:
Cache created library handles.

This ensures newly-created tasks don't have to spend time on
creating these handles (which often requires memory allocations
and/or global synchronization) during their first library call.

maleadt committed:
Key the cublasXt handle against all contexts.

maleadt committed:
Update the stream of libraries eagerly.

This makes stream switching take 500ns instead of 80, but it simplifies handle management
(making sure any active handles immediately use the new stream, reducing the risk of a
mismatch there), and avoids needless stream switches when switching tasks. The latter
seems like it might happen much more frequently, so it makes sense to have the overhead
when switching instead of when querying.

Furthermore, for KA.jl many APIs take stream arguments again, so it won't have to switch
streams globally anyway.

maleadt committed:
Streamline state management APIs.

maleadt committed:
Test task-bound streams.

maleadt committed:
Exception tests: catch launch failures.

Now that we don't check for errors on _every_ API call,
it's possible an exception doesn't get caught by
CUDA.check_exceptions, but results in a CUDA error instead.

maleadt committed:
Also check for exceptions during synchronize_all.

maleadt committed:
Simplify context initialization for driver API wrappers.

maleadt committed:
Fix missed detection of device switches.

This state management is too tricky.
